src/main/java/com/google/devtools/build/lib/util/StringEncoding.java

```diff
@@ -0,0 +1,183 @@
+// limitations under the License.
+
+package com.google.devtools.build.lib.util;
+
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.google.devtools.build.lib.unsafe.StringUnsafe;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+
+/**
+ * Utility functions for reencoding strings between Bazel's internal raw byte encoding and regular
+ * Java strings.
+ *
+ * <p>Bazel needs to support the following two setups:
+ *
+ * <ul>
+ *   <li>File paths, command-line arguments, environment variables, BUILD and .bzl files are all
+ *       encoded in UTF-8, on Linux, macOS or Windows.
+ *   <li>File paths, command-line arguments, environment variables, BUILD and .bzl files are all
+ *       encoded in <i>some</i> consistent encoding, on Linux and with the en_US.ISO-8859-1 locale
+ *       available on the host (legacy setup). In particular, this setup allows any byte sequence to
+ *       appear in a file path and be referenced in a BUILD file.
+ * </ul>
+ *
+ * <p>Bazel achieves this by forcing an en_US.ISO-8859-1 locale on Unix when available, which due to
+ * the byte-based nature of Unix APIs allows all Java (N)IO functions to treat strings as raw byte
+ * sequences (a Latin-1 character is equivalent to an unconstrained byte value). On macOS, where the
+ * JVM forces UTF-8 encoding for any kind of system interaction, as well as on Windows, where system
+ * APIs are all restricted to valid Unicode strings, Bazel has to reencode strings to Unicode before
+ * passing them to the JVM (and vice versa). Since BUILD and .bzl files are always read into Latin-1
+ * strings (file encodings are not forced by the JVM) and are assumed to be encoded in UTF-8 (unless
+ * the Latin-1 locale is available), Bazel has to reencode the strings to UTF-8 so that they match
+ * up with the Starlark contents of these files (e.g. file paths mentioned in a BUILD file).
+ *
+ * <p>While allowing the user a great deal of flexibility, this requires great care when {@link
+ * String}s are passed into or out of Bazel via Java standard library functions or external APIs.
+ * The following three difference types of strings need to be distinguished as if they were
+ * different Java types:
+ *
+ * <ul>
+ *   <li>Internal strings: All strings retained by Bazel and used in its inner layers are expected
+ *       to be raw byte sequences stored in Latin-1 {@link String}s. With Java's compact string
+ *       representation, this means that the Latin-1 bytes are stored directly in the internal byte
+ *       array {@link String#value} and the {@link String#coder} is {@link String#LATIN1}.
+ *   <li>Unicode strings: Regular Java strings, which are always Unicode. A common example is a
+ *       {@code string} field in a protobuf message.
+ *   <li>Platform strings: Strings that are passed to or returned from Java (N)IO functions or as
```

Comment:
How are platform strings different from Unicode strings?

AFAIU Java Strings don't have a user-visible "encoding" value so I maybe you mean the encoding of their internal byte array? But that can't be cause you just use `new String()` in `internalToPlatform()` below just like in `internalToUnicode()`.

The distinction makes sense philosophically, but I thought the practical difference would only matter when transforming these String objects into byte arrays directly before calling the APIs of the operating system.

Reply:
The only difference is the condition used to decide whether to reencode: platform strings may be Latin-1 arrays of raw bytes that we want to preserve as such, Unicode strings aren't.